### PR TITLE
ENH: use derivatives for boxcox transformation parameter and error estimation

### PR DESCRIPTION
**Pull request by argriffing**
This uses tricks from http://jekyll.math.byuh.edu/papers/mspaper99.pdf as suggested in https://github.com/statsmodels/statsmodels/issues/1309 for point estimates and error interval estimates of the lambda parameter that box-cox uses for transformation to normality. The idea is to use first and second derivatives of boxcox_llf to search more intelligently for the maximum likelihood estimate, and to use the second derivative of boxcox_llf along with asymptotic properties of the MLE to estimate an interval.

Work in progress includes updating docstrings, in particular the doctests and explaining the deal with the hypergeometric functions. The purpose of sending the PR early is to get feedback on interface changes, and argument names, etc.

### CODE REVIEW

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -861,6 +880,11 @@ def boxcox(x, lmbda=None, alpha=None):
         If ``alpha`` is not None, return the ``100 * (1-alpha)%`` confidence
         interval for `lmbda` as the third output argument.
         Must be between 0.0 and 1.0.
+    lmbda_estimation : str, optional
+        One of {'pearsonr', 'mle', 'mle-newton'}.
+    conf_method : str, optional
+        One of {'lrt', 'asy'}.  This is only used if ``alpha`` is not None.
+        Default: 'lrt'
```
Comment (larsmans) [inline scipy/stats/morestats.py]: It would be nice to have either explanations, or less strings (lrt? asy?).

File: scipy/stats/morestats.py [hunk 2]
```diff
@@ -937,21 +961,81 @@ def boxcox(x, lmbda=None, alpha=None):
...
         # Find confidence interval
-        interval = _boxcox_conf_interval(x, lmax, alpha)
+        interval = _boxcox_conf_interval(x, lmax, alpha, method=conf_method)
         return y, lmax, interval
 
 
+def _boxcox_llf_derivs(lam, logy):
+    # equations (6, 15)
+    # Note that z, u, v are centered here but not in the paper.
+    # Here v is negated relative to v in the paper.
```
Comment (larsmans) [inline scipy/stats/morestats.py]: There's no full reference for "the paper".

**Comment by argriffing** [comment]
@larsmans Thanks for the comments, sorry I probably should not have sent this PR in such preliminary form.

My main question is about how the interface should work, in a way that maintains backwards compatibility. The contribution of this PR consists of two things:
1) A new way to compute the maximum likelihood point estimate of lambda. This new technique has a high convergence order but is perhaps less robust to bad initial guesses.  It is strictly a numerical contribution; mathematically and statistically it tries to compute the same maximum likelihood estimate as in the existing implementation.
2) A new kind of confidence interval for the maximum likelihood point estimate. This uses the [asymptotic normality of the MLE](https://en.wikipedia.org/wiki/Maximum_likelihood#Asymptotic_normality) and simply reports the confidence interval corresponding to this normal approximation.  This interval is constructed in a different way, using different statistical assumptions, than the existing [deviance](https://en.wikipedia.org/wiki/Deviance_(statistics))-based strategy.  It is computed using a "closed form" as opposed to a more open-ended search that is used to compute the deviance-based interval.

These "second thing" contributions neither directly replace nor add to lists of existing options, so the interface might be messy. By "second thing" I mean adding a second maximum likelihood estimation strategy, and adding a second interval definition. Right now I'm kind of abusing the list of estimators by adding 'mle-newton' even though there is an existing 'mle' and both estimators are mathematically the same.  For the interval, I'm adding a new keyword argument with 'lrt' and 'asy' options.  I'm not sure at all about these decisions.  I'm not sure that adding keywords is the best way to do this, and even if keywords should be added I don't have much of an opinion about their names.

**Comment by argriffing** [comment]
One of the boxcox lambda approximation strategies uses the Filliben approximation which I'd like to expose as a public function in https://github.com/scipy/scipy/pull/5124.

**Comment by argriffing** [comment]
closing https://mail.scipy.org/pipermail/numpy-discussion/2016-January/074810.html

**Pull request closed without merging**